File: app/src/main/java/com/ufkoku/ciphersharedpreferences/demo/ui/Adapter.java
package com.ufkoku.ciphersharedpreferences.demo.ui;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.ufkoku.ciphersharedpreferences.R;
import com.ufkoku.ciphersharedpreferences.demo.entity.PrefsEntry;

import java.util.ArrayList;
import java.util.List;

public class Adapter extends RecyclerView.Adapter<Adapter.ItemViewHolder> {

    private Listener listener;

    private LayoutInflater inflater;
    private List<PrefsEntry> entries = new ArrayList<>();

    public Adapter(LayoutInflater inflater) {
        this.inflater = inflater;
    }

    public Adapter(LayoutInflater inflater, List<PrefsEntry> entries) {
        this.inflater = inflater;
        this.entries.addAll(entries);
    }

    public void setListener(Listener listener) {
        this.listener = listener;
    }

    public void addEntry(PrefsEntry entry) {
        entries.add(entry);
        notifyItemInserted(entries.size() - 1);
    }

    public void addEntries(List<PrefsEntry> entriesToAdd) {
        int oldSize = entries.size();
        entries.addAll(entriesToAdd);
        notifyItemRangeInserted(oldSize, entriesToAdd.size());
    }

    public void removeEntry(PrefsEntry entry){
        int index = entries.indexOf(entry);
        if (index != -1){
            entries.remove(index);
            notifyItemRemoved(index);
        }
    }

    @Override
    public ItemViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new ItemViewHolder(inflater.inflate(R.layout.list_item, parent, false));
    }

    @Override
    public void onBindViewHolder(ItemViewHolder holder, int position) {
        holder.bindObject(entries.get(position));
    }

    @Override
    public int getItemCount() {
        return entries.size();
    }

    public class ItemViewHolder extends RecyclerView.ViewHolder {

        protected PrefsEntry binded;

        protected TextView tvKey;
        protected TextView tvValue;
        protected TextView tvEncryptedValue;

        public ItemViewHolder(View itemView) {
            super(itemView);

            tvKey = (TextView) itemView.findViewById(R.id.key);
            tvValue = (TextView) itemView.findViewById(R.id.value);
            tvEncryptedValue = (TextView) itemView.findViewById(R.id.valueEncrypted);

            itemView.findViewById(R.id.deleteButton).setOnClickListener(v -> {
                if (listener != null && binded != null) {
                    listener.onDeleteClicked(binded);
                }
            });
        }

        protected void bindObject(PrefsEntry entry) {
            binded = entry;

            tvKey.setText(entry.getKey());
            tvValue.setText(entry.getValue());
            tvEncryptedValue.setText(entry.getEncryptedValue());
        }

    }

    public interface Listener {

        void onDeleteClicked(PrefsEntry entry);

    }

}

File: app/src/main/kotlin/com/ufkoku/ciphersharedpreferences/demo/ui/main/Adapter.kt
package com.ufkoku.ciphersharedpreferences.demo.ui.main

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.ufkoku.ciphersharedpreferences.R
import com.ufkoku.ciphersharedpreferences.demo.entity.PrefsEntry
import java.util.*

class Adapter : RecyclerView.Adapter<Adapter.ItemViewHolder> {

    var listener: Listener? = null

    private val inflater: LayoutInflater

    private val entries = ArrayList<PrefsEntry>()

    constructor(inflater: LayoutInflater) {
        this.inflater = inflater
    }

    constructor(inflater: LayoutInflater, entries: List<PrefsEntry>) : this(inflater) {
        this.entries.addAll(entries)
    }

    fun addEntry(entry: PrefsEntry) {
        entries.add(entry)
        notifyItemInserted(entries.size - 1)
    }

    fun addEntries(entriesToAdd: List<PrefsEntry>) {
        val oldSize = entries.size
        entries.addAll(entriesToAdd)
        notifyItemRangeInserted(oldSize, entriesToAdd.size)
    }

    fun removeEntry(entry: PrefsEntry) {
        val index = entries.indexOf(entry)
        if (index != -1) {
            entries.removeAt(index)
            notifyItemRemoved(index)
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ItemViewHolder {
        return ItemViewHolder(inflater!!.inflate(R.layout.list_item, parent, false))
    }

    override fun onBindViewHolder(holder: ItemViewHolder, position: Int) {
        holder.bindObject(entries[position])
    }

    override fun getItemCount(): Int {
        return entries.size
    }

    inner class ItemViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        protected var binded: PrefsEntry? = null

        protected val tvKey: TextView = itemView.findViewById(R.id.key)

        protected val tvValue: TextView = itemView.findViewById(R.id.value)

        protected val tvEncryptedValue: TextView = itemView.findViewById(R.id.valueEncrypted)

        init {
            itemView.findViewById<View>(R.id.deleteButton).setOnClickListener {
                if (listener != null && binded != null) {
                    listener!!.onDeleteClicked(binded!!)
                }
            }
        }

        fun bindObject(entry: PrefsEntry) {
            binded = entry

            tvKey.text = entry.key
            tvValue.text = entry.value
            tvEncryptedValue.text = entry.encryptedValue
        }

    }

    interface Listener {

        fun onDeleteClicked(entry: PrefsEntry)

    }

}
